core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/common/ReachableValues.java

```diff
@@ -157,59 +186,75 @@ boolean containsValue(int valueIndex) {
             return valueBitSet.get(valueIndex);
         }
 
-        private static int[] extractAllIndexes(BitSet bitSet) {
-            var indexes = new int[bitSet.cardinality()];
-            var idx = 0;
+        private static List<Integer> extractAllIndexes(BitSet bitSet) {
+            var indexes = new ArrayList<Integer>(bitSet.cardinality());
             for (int i = bitSet.nextSetBit(0); i >= 0; i = bitSet.nextSetBit(i + 1)) {
-                indexes[idx++] = i;
+                indexes.add(i);
             }
             return indexes;
         }
 
         List<Object> getRandomAccessEntityList(List<Object> allEntities) {
-            if (onDemandRandomAccessEntityList == null) {
-                onDemandRandomAccessEntityList = new ArrayIndexedList<>(extractAllIndexes(entityBitSet), allEntities, null);
-            }
-            return onDemandRandomAccessEntityList;
+            return new ArrayIndexedList<>(extractAllIndexes(entityBitSet), allEntities, null);
         }
 
-        List<Object> getRandomAccessValueList(List<ReachableItemValue> allValues) {
-            if (onDemandRandomAccessValueList == null) {
-                onDemandRandomAccessValueList = new ArrayIndexedList<>(extractAllIndexes(valueBitSet), allValues, v -> v.value);
+        <V> List<Object> getRandomAccessValueList(List<ReachableItemValue> allValues,
+                @Nullable ValueRangeSorter<V> valueRangeSorter) {
+            var valuesList = new ArrayIndexedList<>(extractAllIndexes(valueBitSet), allValues,
+                    v -> (V) v.value);
+            if (valueRangeSorter != null) {
+                valueRangeSorter.sort(valuesList);
             }
-            return onDemandRandomAccessValueList;
+            return (List<Object>) valuesList;
         }
     }
 
+    @NullMarked
+    public record ReachableValuesIndex<T>(Map<Object, Integer> indexMap, List<T> allItems) {
```

Comment:
Arguably same here. What is the object here? Is it a value? Is it an entity? Can it be both? The generic should say that.